enhancements/20220530-vxlan-enhancement.md

```diff
@@ -0,0 +1,521 @@
+
+Advantage:
+  
+  Inter-subnet traffic can be centrally managed.
+  
+  Gateway deployment and management are simple.
+
+Disadvantages:
+
+  Forwarding paths are not optimal. Inter-subnet Layer 3 traffic of the same Layer 2 gateway must be transmitted to the centralized Layer 3 gateway for forwarding.
+
+  The ARP entry may be too big, in worst cases, all VM's MAC are in there. (VIP can reduce the number)
+
+##### Distributed Gateway
+
+  A big topic, BGP EVPN and more are included.
+
+  TBD
+
+#### Story 7
+
+VM-internet communication
+
+For VM Group or Guest Kubernetes Cluster, they need to communicate with internet and vice versa.
+
+It is described together in story 9.
+
+#### Story 8
+
+Guest Kubernetes Cluster - Rancher Integration
+
+##### Guest Kubernetes Cluster creation
+
+https://docs.harvesterhci.io/v1.0/rancher/rke2-cluster/
+
+When creating an RKE2 Kubernetes Cluster via Harvester Node Driver, user select the reateld vxlan_id/vni_id . (vni may be created in advance or automatically)
+
+![](./20220530-vxlan-enhancement/guest-k8s-cluster-create-1.png)
+
+##### VM Management
+
+According to following document, the VM management is done in Harvester UI. Rancher document does not mention this part.
+
+https://docs.harvesterhci.io/v1.0/rancher/virtualization-management/
+
+Open Points:
+
+1. When a new VM is created for a specific Guest Kubernetes Cluster in Harvester, will it be added to this Cluster automatically ?
+
+2. When VM has no eth0 (management network) attached, can Harvester UI use WebRNC to connect it?
+
+##### Guest Kubernetes Cluster management
+
+Per Rancher document, Rancher will contact with Guest Kubernetes Cluster via CLI or RancherUI, it means, Rancher needs to be network reachable with Guest Kubernetes Cluster, at least it's VIP.
+
+https://rancher.com/docs/rancher/v2.6/en/cluster-admin/cluster-access/
+
+https://rancher.com/docs/rancher/v2.6/en/cluster-admin/cluster-access/kubectl/
+
+![](./20220530-vxlan-enhancement/rancher-integration-1.png)
+
+
+#### Story 9
+
+Guest Kubernetes Cluster VIP/LP and communication with outside
+
+
+##### How VLAN network handle
+
+It is well described here.
+
+https://github.com/yaocw2020/harvester/blob/ccm-HEP/enhancements/20220214-harvester-cloud-provider-enhancement.md
+
+![](./20220530-vxlan-enhancement/vlan-lb-1.png)
+
+The Guest Kubernetes Cluster VIP is from provider network, it is reachable to other systems (like Rancher) in provider network.
+
+note:
+```
+The step 8, gratuitous ARP, which is a key step for kube-vip to float the VIP.
+```
+
+##### Centralized NAT
+
+VXLAN provides overlay network, namely, the Guest Kubernetes Cluster IPs in not reachable from provider network by default.
+
+One way is to have NAT in L3 VXLAN Gateway.
+
+###### NAT 1:1 IP mapping
+
+To hide the details of overlay network, the VIP/LB of Guest Kubernetes Cluster, have a 1:1 mapping public IP/Host Cluster IP.
```

Comment:
We should not use LB to do the function of EIP.
LB will do SNAT to modify packet and make packet sent from LB to VM.
EIP will keep the original client info in L3 packet.